File: 2197-DecodeTheSlantedCiphertext/2197-DecodeTheSlantedCiphertext.java
// Last updated: 07/04/2026, 11:35:17
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        
        int n = encodedText.length();
        if (n == 0) return "";
        
        int cols = n / rows;

        // Step 1: Fill matrix
        char[][] M = new char[rows][cols];
        int k = 0;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                M[r][c] = encodedText.charAt(k++);
            }
        }

        // Step 2: Traverse diagonally
        StringBuilder ans = new StringBuilder();

        for (int c = 0; c < cols; c++) {
            int r = 0;
            int col = c;

            while (r < rows && col < cols) {
                ans.append(M[r++][col++]);
            }
        }

        // Step 3: Remove trailing spaces
        int end = ans.length() - 1;
        while (end >= 0 && ans.charAt(end) == ' ') {
            end--;
        }

        return ans.substring(0, end + 1);
    }
}

File: 2197-DecodeTheSlantedCiphertext/2197-DecodeTheSlantedCiphertext.cpp
// Last updated: 07/04/2026, 11:35:22
class Solution {
public:
    string decodeCiphertext(string encodedText, int rows) {
        string ans ="";
        
        int rc = encodedText.size();
        int r = rows;
        int c = rc/r;

        vector<vector<char>> M(r,vector<char>(c,0));
        int k = 0;

        for (int i = 0 ;i<r;i++){
            for (int j = 0;j<c;j++){
                
                M[i][j] =encodedText[k++];
            }
            if ( k == rc ) break;
        }


        for ( int col = 0;col<c;col++){

            int row = 0;
            int k = col;
            while(row <r && k <c){
                
                ans+=M[row++][k++];
                
            }
        }


        int n = ans.size();
        int m = n;
        for ( int i = n-1;i>=0;i--){
            if ( ans[i] ==' ') m--;
            if ( ans[i]!=' ')break;
        }
        return ans.substr(0,m);

        
       
    }
};